C7/PCXToGodot.cs

```diff
@@ -5,65 +5,102 @@
 public class PCXToGodot : Godot.Object
 {
 	private readonly static byte CIV3_TRANSPARENCY_START = 254;
-	
-	public static ImageTexture getImageTextureFromPCX(Pcx pcx) {
-		Image ImgTxtr = ByteArrayToImage(pcx.ColorIndices, pcx.Palette, pcx.Width, pcx.Height);
+
+	public static ImageTexture getImageTextureFromImage(Image image) {
 		ImageTexture Txtr = new ImageTexture();
-		Txtr.CreateFromImage(ImgTxtr, 0);
+		Txtr.CreateFromImage(image, 0);
 		return Txtr;
 	}
 
+	public static ImageTexture getImageTextureFromPCX(Pcx pcx) {
+		Image ImgTxtr = ByteArrayToImage(pcx.ColorIndices, pcx.Palette, pcx.Width, pcx.Height);
+		return getImageTextureFromImage(ImgTxtr);
+	}
+
 	public static ImageTexture getImageTextureFromPCX(Pcx pcx, int leftStart, int topStart, int croppedWidth, int croppedHeight) {
-		Image Image = getImageFromPCX(pcx, leftStart, topStart, croppedWidth, croppedHeight);
-		ImageTexture Txtr = new ImageTexture();
-		Txtr.CreateFromImage(Image, 0);
-		return Txtr;
+		Image image = getImageFromPCX(pcx, leftStart, topStart, croppedWidth, croppedHeight);
+		return getImageTextureFromImage(image);
+	}
+
+	// BufferData should be as large as the largest PCX file needed to be loaded in pixels
+	// For now, I will assume no PCX file is larger than 4k resolution
+	// It's efficient to use a shared BufferData, ColorData, and AlphaData instead of having to reinitialize these arrays for every Image load
+	// However, this will become a problem if (when?) we move towards parallelization
+	// But for now, it's an improvement
+	public static int[] BufferData = new int[3840 * 2160];
+	public static int[] ColorData = new int[256];
+	public static int[] AlphaData = new int[256];
```

Comment:
Is there a measurable improvement to not allocating 8.5 MB or so of data?  To me this seems like a micro-optimization that will burn someone when they unsuspectingly try to open two PCX files at once (granted, via threads, but at some point someone will likely try it).  And since most PCX files are a lot smaller than 4K, it's a lot less than 8 MP per file... pollution.pcx would only be saving a bit over 200KB of allocation.

If there's data to show that this makes a big difference, maybe... but even then there should be some locks around the entry points to avoid this being a footgun.  And as one of the goals is to be more stable and have fewer limitations than Civ3, I'd rather err on the side of safety.

(Dynamic buffer data allocation would also cover that inevitable 8K texture from 2032... it may seem far-fetched but I've already seen a full-screen texture in a Civ3 mod from China, so I wouldn't be shocked if someone with one of those lovely 5K LG or Dell monitors tries something like that)

Reply:
My gut is that of all the changes commented on so far, this is the one that will cause the biggest performance drop if reverted. However, I need to actually refactor these static variables back into local ones, which is a change I'm a bit too sleepy to test out right now. And yeah, it's not great in terms of future expectations, so maybe the old approach would be better here.